packages/runtime/container-runtime/src/opLifecycle/opGroupingManager.ts

```diff
@@ -0,0 +1,71 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+import { ContainerRuntimeMessage } from "..";
+import { IBatch } from "./definitions";
+
+interface IGroupedMessage {
+	contents: string;
+	metadata: Record<string, unknown> | undefined;
+	compression: string;
+}
+
+export class OpGroupingManager {
+	static groupedBatchOp = "groupedBatch";
+
+	constructor(private readonly groupedBatchingEnabled: boolean) {}
+
+	public groupBatch(batch: IBatch): IBatch {
+		if (batch.content.length < 2 || !this.groupedBatchingEnabled) {
+			return batch;
+		}
+
+		// Need deserializedContent for back-compat
+		const deserializedContent = {
+			type: OpGroupingManager.groupedBatchOp,
+			contents: batch.content.map((message) => ({
+				contents: message.contents,
+				metadata: message.metadata,
+				compression: message.compression,
+			})),
+		};
+
+		const groupedBatch: IBatch = {
+			...batch,
+			content: [
+				{
+					...batch.content[0],
+					metadata: undefined, // important to not keep reference to first op's batch metadata
+					compression: undefined,
+					deserializedContent: deserializedContent as ContainerRuntimeMessage,
```

Comment:
deserializedContent - I think your other PR removed tracking it, no?

Reply:
That change was in next and just missed the cutoff https://github.com/microsoft/FluidFramework/pull/14462